src/Microsoft.OData.Core/ODataStream.cs

```diff
@@ -50,20 +62,53 @@ public override long Seek(long offset, SeekOrigin origin)
         /// <param name="disposing">True if called from Dispose; false if called form the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
-                if (this.listener != null)
+                // Tell the listener that the stream is being disposed; we expect
+                // that no asynchronous action is triggered by doing so.
+                if (this.synchronous)
                 {
-                    // Tell the listener that the stream is being disposed; we expect
-                    // that no asynchronous action is triggered by doing so.
-                    this.listener.StreamDisposed();
-                    this.listener = null;
+                    this.listener?.StreamDisposed();
                 }
+                else
+                {
+                    this.listener?.StreamDisposedAsync().Wait();
+                }
+
+                this.listener = null;
             }
 
+            this.disposed = true;
             base.Dispose(disposing);
         }
 
+#if NETSTANDARD2_0
+        public async ValueTask DisposeAsync()
+        {
+            await DisposeAsyncCore()
+                .ConfigureAwait(false);
+
+            // Dispose unmanaged resources
+            // Pass `false` to ensure functional equivalence with the synchronous dispose pattern
+            this.Dispose(false);
+        }
+
+        /// <summary>
+        /// Encapsulates the common asynchronous cleanup operations.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        protected virtual async ValueTask DisposeAsyncCore()
+        {
+            if (!this.disposed && this.listener != null)
+            {
+                await this.listener.StreamDisposedAsync()
+                    .ConfigureAwait(false);
+
+                this.listener = null;

```

Comment:
In the async version you never set this.disposed=true like you do in the sync version. You do set listener to false, which ensures it won't be executed again (although the synchronous dispose would pass the check on line 65 and then be a no-op).

So I think it all works, I'm just wondering why you rely on both disposed=false and listener!=null and handle them differently between sync and async.

Reply:
@mikepizzo `DisposeAsyncCore` is called from `DisposeAsync` and following the [guidance from Task Asynchronous Programming team](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns), `DisposeAsync` in turn calls `Dispose` method, passing `false` as a parameter to ensure functional equivalency with the synchronous pattern. `Dispose` method in turn sets `disposed` to `true` and that way idempotency is guaranteed.

Reply:
I missed that DisposeAsync calls Dispose -- makes sense, thanks John.